Add Husqvarna Automower integration

Thomas55555 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add Husqvarna Automower integration. This was previously a custom integration. As there is the `lawn_mower` entity now, I thought it's a good moment to integrate it in Home Assistant.
Replaces messed up  #99776

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/28817

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

robinostlund commented:
@Thomas55555 seems to be some issues with the pytests

Thomas55555 commented:
I've improved the tests, but they are not passing yet, because I can't handle this in the tests:

https://github.com/home-assistant/core/blob/ddf94cd570b0fdf66ccda9f9c1ce860bfbe09c12/homeassistant/components/husqvarna_automower/__init__.py#L29-L33

In the previous PR, we focused on getting the code right, and the tests were blocked in the CI.
How will it be now, should I also finish the tests, or can I do this later, after I got an okay for the normal code?

robinostlund commented:
@Thomas55555 just a question, this integration will miss the calendar and all the other sensors right?

MartinHjelmare commented:
We need to finish the tests before merging the PR, but it's ok to wait with the tests until the rest of the code is done.

MartinHjelmare commented on homeassistant/components/husqvarna_automower/__init__.py:
```diff
@@ -0,0 +1,65 @@
+    """Handle unload of an entry."""
+    coordinator = hass.data[DOMAIN][entry.entry_id]
+    with contextlib.suppress(Exception):
+        await coordinator.session.close()
```
We also need to clean up running tasks on Home Assistant stop event. Search for `EVENT_HOMEASSISTANT_STOP` for examples.

MartinHjelmare commented on tests/components/husqvarna_automower/test_application_credentials.py:
```diff
@@ -0,0 +1,72 @@
+from tests.common import MockConfigEntry
+
+
+@pytest.mark.asyncio
```
```suggestion
```

MartinHjelmare commented on tests/components/husqvarna_automower/test_application_credentials.py:
```diff
@@ -0,0 +1,72 @@
+    return config_entry
+
+
+@pytest.mark.asyncio
```
```suggestion
```

MartinHjelmare commented on tests/components/husqvarna_automower/test_common.py:
```diff
@@ -0,0 +1,187 @@
+
+
+@dataclass
+class System:
```
These dataclasses should not be in Home Assistant but in the library.

MartinHjelmare commented on tests/components/husqvarna_automower/test_lawn_mower.py:
```diff
@@ -0,0 +1,169 @@
+        mower.data[0].attributes.mower.activity = activity
+
+    assert lawn_mower._attr_unique_id == f"{MWR_ONE_ID}_lawn_mower"
+    set_activity("")
```
Why do we set it to empty string? I'd not use empty string to mean something in the library.

Thomas55555 commented:
> question, this integration will be missing the calendar and all the other sensors right?

In the beginning yes. Later I want to add additional platforms.

Thomas55555 commented on homeassistant/components/husqvarna_automower/config_flow.py:
```diff
@@ -0,0 +1,36 @@
+
+    async def async_oauth_create_entry(self, data: dict[str, Any]) -> FlowResult:
+        """Create an entry for the flow."""
+        user_id = data.get(CONF_TOKEN, {}).get("user_id")
```
No, it's not optional. I changed it.

Thomas55555 commented on homeassistant/components/husqvarna_automower/config_flow.py:
```diff
@@ -0,0 +1,36 @@
+        await self.async_set_unique_id(user_id)
+        self._abort_if_unique_id_configured()
+        return self.async_create_entry(
+            title=NAME,
```
It's not user friendly. But we can get some user information out of the JWT and make an individual for each config entry. See:
https://github.com/home-assistant/core/pull/100569/commits/db860afa5151ce715d24b2397c2f4034193be510

MartinHjelmare commented on homeassistant/components/husqvarna_automower/coordinator.py:
```diff
@@ -26,14 +28,32 @@ def __init__(
+        if self.data:
+            for mower in self.data.data:
+                if mower.id == ws_data.id:
+                    for field in fields(ws_data.attributes):
```
I suggest making a helper method on the `mower` class in the library that takes `ws_data` and updates the mower attributes.

Thomas55555 replied:
But then I would have to pass the polled data from the coordinator to the library. Wouldn't that be too complicated, without a real benefit?

MartinHjelmare replied:
No, I suggest we pass the websocket update data, ie the `MowerData` instance  to a method on the existing stored `MowerData` instance. The benefit is that the library takes care of the details of how to apply the updated websocket data.

Alternatively since the library already stores the `MowerData` instances on the client, the websocket could update them directly and just let the library user know which mower was updated.

Also, I suggest storing the `MowerData` instances in a dict instead of a list, if the `id` of the mower data instance is unique. That way we can easily look up a specific mower data instance if we have the id.

Thomas55555 replied:
Okay, got it.

MartinHjelmare commented on homeassistant/components/husqvarna_automower/coordinator.py:
```diff
@@ -37,19 +37,22 @@ def __init__(
-        )
+        if not self.ws_connected:
+            await self.mowersession.connect()
+            self.mowersession.register_data_callback(
```
If this just stores the callback reference, we probably don't need to do it after every connect call, but just once for the coordinator. Or is the callback unregistered if the connection is lost?

Thomas55555 commented on homeassistant/components/husqvarna_automower/api.py:
```diff
@@ -0,0 +1,26 @@
+from homeassistant.helpers import config_entry_oauth2_flow
+
+
+class AsyncConfigEntryAuth(aioautomower.AbstractAuth):
```
I've reworked the library according to the documentation. I actually like this approach.
Everything works fine now except of the websocket connection. It's not possible to establish it and I don't get an error. I've tried very long, but wasn't able to get it working. Could you have a look at the library, please?
https://github.com/Thomas55555/aioautomower/blob/a196d9fbbff1f9fddf65a3eed3d029872ce51a00/aioautomower/session.py#L371

MartinHjelmare replied:
I don't think you should make a loop when connecting the websocket. Just have a `connect` method that connects the websocket and stores a reference to it on the client.

https://github.com/Thomas55555/aioautomower/blob/a196d9fbbff1f9fddf65a3eed3d029872ce51a00/aioautomower/auth.py#L148-L160

Then have another method `listen` that fetches incoming messages and iterates over them and handles them. It's this method that the library user should create a background task for.

https://github.com/Thomas55555/aioautomower/blob/a196d9fbbff1f9fddf65a3eed3d029872ce51a00/aioautomower/session.py#L374

The benefit of separating the `connect` and `listen` methods is that we can make sure we can connect the websocket with a timeout before we create the task that should listen for websocket messages. If we can connect the websocket the chances are high that the websocket communication will work. If we fail to connect during the timeout, we want to raise `ConfigEntryNotReady` and try again later automatically.

Side notes:
1. It looks like there's a double call to update data callbacks:
https://github.com/Thomas55555/aioautomower/blob/a196d9fbbff1f9fddf65a3eed3d029872ce51a00/aioautomower/session.py#L380-381
https://github.com/Thomas55555/aioautomower/blob/a196d9fbbff1f9fddf65a3eed3d029872ce51a00/aioautomower/session.py#L342

2. I'd remove the schedule immediately parameter. The library user can call the callback themselves if needed.
https://github.com/Thomas55555/aioautomower/blob/a196d9fbbff1f9fddf65a3eed3d029872ce51a00/aioautomower/session.py#L74

MartinHjelmare replied:
If we create a `connect` method in the library for the websocket, there should also be a `disconnect` method for the websocket.

Thomas55555 replied:
There is now a `close` method

MartinHjelmare commented on homeassistant/components/husqvarna_automower/lawn_mower.py:
```diff
@@ -98,12 +99,21 @@ def activity(self) -> LawnMowerActivity:
+        try:
+            await self.coordinator.api.resume_schedule(self.mower_id)
+        except ApiException as exception:
+            _LOGGER.error("Command couldn't be sent to the command que: %s", exception)
```
Please raise `HomeAssistantError` with the message as exception argument instead of logging the message.

The message will show in the frontend.

https://developers.home-assistant.io/docs/integration_quality_scale_index#silver-

MartinHjelmare commented on tests/components/husqvarna_automower/const.py:
```diff
@@ -0,0 +1,226 @@
+MWR_TWO_IDX = 1
+
+
+MOWER_ONE_SESSION_DATA = {
```
The large data constants can be stored as JSON files in a `fixtures` directory under this integration tests. We have helpers to load fixture files.

https://github.com/home-assistant/core/blob/4ae5757bc175307545e45519cb64a1d24b9f763f/tests/common.py#L470-L494

MartinHjelmare commented on tests/components/husqvarna_automower/test_lawn_mower.py:
```diff
@@ -61,109 +101,97 @@ async def setup_entity(hass: HomeAssistant):
     await setup_entity(hass)
-    coordinator = hass.data[DOMAIN]["automower_test"]
-    lawn_mower = AutomowerLawnMowerEntity(mower.data[0], coordinator)
+    coordinator: AutomowerDataUpdateCoordinator = hass.data[DOMAIN]["automower_test"]
```
Please don't interact directly with the coordinator or any entities. Those are integrations details. Set up the integration while patching the client library appropriately and assert core state.

https://developers.home-assistant.io/docs/development_testing#writing-tests-for-integrations

Thomas55555 replied:
I've changed some of the tests in this file. Can you tell me, if this is going in the right direction?

MartinHjelmare replied:
Yes, I added some comments below.

MartinHjelmare commented:
I hoping we can finish this PR at some point.

Thomas55555 commented:
I was quite busy the last months. But I was working on the library the last days.
I also hope we can finish it soon.

MartinHjelmare commented on tests/components/husqvarna_automower/test_lawn_mower.py:
```diff
@@ -0,0 +1,133 @@
+            resume_schedule=AsyncMock(),
+        ),
+    ), patch(
+        "homeassistant.helpers.config_entry_oauth2_flow.async_get_config_entry_implementation",
```
We shouldn't need to patch all these details.

Thomas55555 replied:
If it isn't patched, there are errors like that: `FAILED tests/components/husqvarna_automower/test_lawn_mower.py::test_lawn_mower_commands - TypeError: <AsyncMock name='async_get_config_entry_implementation().async_refresh_token()' id='139843651950352'>`
I don't know another way to avoid them.

MartinHjelmare commented on tests/components/husqvarna_automower/test_lawn_mower.py:
```diff
@@ -0,0 +1,133 @@
+TEST_MOWER_ID = "c7233734-b219-4287-a173-08e3643f89f0"
+
+
+@pytest.mark.enable_socket
```
Why are we enabling the socket?

MartinHjelmare commented on tests/components/husqvarna_automower/test_lawn_mower.py:
```diff
@@ -0,0 +1,133 @@
+
+
+@pytest.mark.enable_socket
+@pytest.mark.asyncio
```
```suggestion
```

MartinHjelmare commented on tests/components/husqvarna_automower/test_lawn_mower.py:
```diff
@@ -0,0 +1,133 @@
+    return config_entry
+
+
+@pytest.mark.asyncio
```
Remove all these.
```suggestion
```

MartinHjelmare commented on tests/components/husqvarna_automower/test_lawn_mower.py:
```diff
@@ -0,0 +1,133 @@
+    """Test lawn_mower state."""
+    mower = load_json_value_fixture("mower.json", DOMAIN)
+    mowers: MowerList = make_mower_list(mower)
+    await setup_entity(hass, mowers)
```
Looks like loading the fixture, making the mowers list and setting up the integration should be three different pytest fixtures. Loading the fixture should be scoped to the test session so its not done more than once.

MartinHjelmare replied:
Here are some examples of pytest fixtures:
https://github.com/home-assistant/core/blob/eb01998395ef091d11dabc66d940ca9c53e3fe70/tests/components/evil_genius_labs/conftest.py#L12-L59

MartinHjelmare reviewed (commented):
The tests are mowing in the right direction. 👍

MartinHjelmare commented on tests/components/husqvarna_automower/conftest.py:
```diff
@@ -0,0 +1,86 @@
+    return load_json_value_fixture("mower.json", DOMAIN)
+
+
+@pytest.fixture(scope="session")
```
I suggest we don't scope this for the session so that we can modify this data freely in each test without worrying about affecting other tests.

```suggestion
@pytest.fixture
```

MartinHjelmare commented on tests/components/husqvarna_automower/conftest.py:
```diff
@@ -0,0 +1,86 @@
+    return mowers
+
+
+@pytest.fixture(scope="session")
```
A small string is probably not worth scoping as session. The cost for setting up the fixture is small. Unless there's some other reason it makes sense to set scope as session?

MartinHjelmare commented on tests/components/husqvarna_automower/conftest.py:
```diff
@@ -0,0 +1,86 @@
+    """Set up entity and config entry."""
+
+    mower_data: MowerAttributes = mower_list_fixture[TEST_MOWER_ID]
+    if activity:
```
Avoid using conditional logic in tests. There should be a single branch. Prefer using `pytest.mark.parametrize` to create different test cases.

A typical refactor here would be to parametrize `activity` so that we can always set it on the mower data.

Thomas55555 replied:
I've removed it.
Parametrized tests are already in use.

Lash-L committed:
Add Translation for Roborock exceptions (#105427)

* add translations to exceptions

* Make errors more user understandable

* make command not ternary operator

* removed non-user facing exceptions

* Add user facing exceptions and code coverage

* add match

* fix linting

* Apply suggestions from code review

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

---------

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

bdraco committed:
Cache commonly called Integration manifest properties (#108141)

joostlek committed:
Delete removed channel devices in Youtube (#107907)

gjohansson-ST committed:
Remove config import in Neato (#107967)

Co-authored-by: jbouwh <jan@jbsoft.nl>

bdraco committed:
Avoid bytes to string to bytes conversion in websocket api (#108139)

gjohansson-ST committed:
Add entity name translations to System Monitor (#107952)

frenck committed:
Deprecate Python 3.11 (#108160)

gjohansson-ST committed:
Remove config import from netatmo (#107972)

* Remove config import from netatmo

* Fix tests

synesthesiam committed:
Wyoming satellite ping and bugfix for local wake word (#108164)

* Refactor with ping

* Fix tests

* Increase test coverage

klaasnicolaas committed:
Set minimal value for modules power - Forecast.solar (#108166)

2franix committed:
Upgrade python-mpd2 to v3.1.1 (#108143)

bdraco committed:
Bump dbus-fast to 2.21.1 (#108176)

bachya committed:
Bump `aiokafka` to 0.10.0 (#108165)

cdce8p committed:
Add decorator typing [yeelight] (#107598)

bachya committed:
Move Guardian coordinator to suggested location (#108182)

* Move Guardian coordinator to suggested location

* Fix coverage

jpbede committed:
Dynamically map state class, device class and UoM in ZHA smart energy metering sensor (#107685)

* Dynamically map state class, device class and UoM in ZHA smart energy metering sensor

* Fix some state & device classes and add scaling

* Fix added imperial gallons tests

* Use entity description instead of custom class & add one entity to tests

* Apply code review suggestion

* Scale only when needed

* Revert "Scale only when needed"

This reverts commit a9e0403402e457248786ff1f231d4b380c2969ce.

* Avoid second lookup of entity description

* Change test to not mix sensor types

bachya committed:
Remove unnecessary OpenUV entity description mixins (#108195)

bachya committed:
Remove unnecessary SimpliSafe entity description mixins (#108197)

bachya committed:
Remove unnecessary PurpleAir entity description mixins (#108196)

bachya committed:
Remove unnecessary Notion entity description mixins (#108194)

bachya committed:
Remove unnecessary AirVisual Pro entity description mixins (#108192)

bachya committed:
Remove unnecessary Ambient PWS entity description mixins (#108191)

bachya committed:
Remove unnecessary RainMachine entity description mixins (#108190)

gjohansson-ST committed:
Create update coordinator for Systemmonitor (#106693)

dependabot[bot] committed:
Bump actions/cache from 3.3.3 to 4.0.0 (#108209)

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

Bre77 committed:
Fix translation keys in Tessie (#108203)

zweckj committed:
Add sensor platform to La Marzocco integration (#108157)

* add sensor

* remove switch

* requested changes

* property instead of function

* add missing snapshot

* rename var, fixture

Moustachauve committed:
Implement TechnoVE integration (#106029)

* Implement TechnoVE integration

Only the basic sensors for now.

* Add technoVE to strict typing

* Implement TechnoVE PR suggestions

* Remove Diagnostic from TechnoVE initial PR

* Switch status sensor to Enum device class

* Revert zeroconf for adding it back in subsequent PR

* Implement changes from feedback in TechnoVE PR

* Update homeassistant/components/technove/models.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* Update homeassistant/components/technove/sensor.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* Update homeassistant/components/technove/models.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* Remove unnecessary translation keys

* Fix existing technoVE tests

* Use snapshot testing for TechnoVE sensors

* Improve unit tests for TechnoVE

* Add missing coverage for technoVE config flow

* Add TechnoVE coordinator tests

* Modify device_fixture for TechnoVE from PR Feedback

* Change CONF_IP_ADDRESS to CONF_HOST for TechnoVE

* Update homeassistant/components/technove/config_flow.py

Co-authored-by: Sid <27780930+autinerd@users.noreply.github.com>

* Update homeassistant/components/technove/models.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* Update homeassistant/components/technove/models.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* Implement feedback from TechnoVE PR

* Add test_sensor_update_failure to TechnoVE sensor tests

* Add test for error recovery during config flow of TechnoVE

* Remove test_coordinator.py from TechnoVE

---------

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>
Co-authored-by: Sid <27780930+autinerd@users.noreply.github.com>

zweckj committed:
Change the way an entity is supported in La Marzocco (#108216)

* refactor supported

* refactor supported

zweckj committed:
Add binary sensor platforms to La Marzocco (#108212)

* add sensor

* remove switch

* requested changes

* property instead of function

* add missing snapshot

* rename var, fixture

* add binary sensors

* reorder strings

* rename sensor

* switch to supported_fn

gjohansson-ST committed:
Allow selecting camera in Trafikverket Camera (#105927)

* Allow selecting camera in Trafikverket Camera

* Final config flow

* Add tests

* Fix load_int

* naming

cdce8p committed:
Improve aurora typing (#108217)

zweckj committed:
Add select platform to La Marzocco integration (#108222)

* add select

* change check, icons

* fix docstrings, use []

zweckj committed:
Add number platform to La Marzocco (#108229)

* add number

* remove key entities

* remove key numbers

* rename entities

* rename sensors

bdraco committed:
Remove unused entity_sources argument from shared_attrs_bytes_from_event (#108210)

frenck committed:
Disable Python 3.12 incompatible integrations (#108163)

zweckj committed:
Add update platform to La Marzocco (#108235)

* add update

* requested changes

* improve

* docstring

* docstring

zweckj committed:
Add button platform to La Marzocco (#108236)

* add button

* Update homeassistant/components/lamarzocco/button.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* Update homeassistant/components/lamarzocco/strings.json

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* update snapshot

---------

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

emontnemery committed:
Assert default response from conversation trigger (#108231)

jbouwh committed:
Remove deprecated redundant dry and fan modes from `zwave_js` climates (#108124)

Remove deprecated redundant dry and fan modes from zwave_js climates

MartinHjelmare committed:
Decrease fitbit logging verbosity on connection error (#108228)

* Add test for connection error

* Decrease fitbit connection error log verbosity

zweckj committed:
Fix state after La Marzocco update (#108244)

request a refresh from coordinator after update

cdce8p committed:
Improve systemmonitor generic typing (#108220)

miaucl committed:
Switch for swiss_public_transport to unique_id instead of unique_entry (#107910)

* use unique_id instead of unique_entry

* move entry mock out of patch context

Thomas55555 committed:
test_init

erwindouna committed:
Bump PyTado to 0.17.4 (#108255)

Bump to 17.4

Commit:
Send target temp to Shelly TRV in F when needed (#108188)

elupus committed:
Let zigpy decide on default manufacturer id (#108257)

* zha: let cluster set default manufacturer id

* zha: allow forcing manufacturer id off

starkillerOG committed:
Bump reolink_aio to 0.8.7 (#108248)

jbouwh committed:
Remove legacy VacuumEntity base class support (#108189)

puddly committed:
Bump ZHA dependency zigpy to 0.60.6 (#108266)

Bump zigpy to 0.60.6

puddly committed:
Speed up ZHA initialization and improve startup responsiveness (#108103)

* Limit concurrency of startup traffic to allow for interactive usage

* Drop `retryable_req`, we already have request retrying

* Oops, `min` -> `max`

* Add a comment describing why `async_initialize` is not concurrent

* Fix existing unit tests

* Break out fetching mains state into its own function to unit test

Kane610 committed:
Bump aiounifi to v69 to improve websocket logging (#108265)

starkillerOG committed:
Use cache update for WIFI blinds (#108224)

mkmer committed:
Allow multiple config entries in Honeywell (#108263)

* Address popping all entires when unloading

* optimize hass data

mkmer committed:
Remove unused variables in honeywell (#108252)

Remove unused configuration keys

bdraco committed:
Fix apple_tv IP Address not being updated from discovery (#107611)

bdraco committed:
Fix benign typo in entity registry (#108270)

dependabot[bot] committed:
Bump github/codeql-action from 3.23.0 to 3.23.1 (#108275)

bdraco committed:
Small cleanups to the websocket api handler (#108274)

bdraco committed:
Avoid many replace calls in find_next_time_expression_time (#108273)

bdraco committed:
Add state caching to button entities (#108272)

cdce8p committed:
Enable strict typing for config (#108023)

cdce8p committed:
Improve daikin typing (#108039)

zweckj committed:
Add diagnostics to La Marzocco (#108240)

* add diagnostics

* make firmware section easier to read

zweckj committed:
Bump lmcloud to 0.4.35 (#108288)

bump lmcloud

lellky committed:
Add tests to flexit_bacnet integration (#108291)

* Add fixture for update method

* Mock flexit_bacnet

* Adds test for climate

* Adds snapshot testing

* Adds test for init, refactor test for config flow

joostlek committed:
Add test for failed initialization in Flexit BACnet (#108294)

lellky committed:
Add coordinator to Flexit bacnet (#108295)

* Adds coordinator and base entity class

* Patch the coordinator

* Adds device property to base class

And refactors accordingly

* Use const instead of string

* Moves _attr_has_entity_name to base entity

* Argument as positional

* Use device_id from init

emontnemery committed:
Don't always set first thread dataset as preferred (#108278)

* Don't always set first thread dataset as preferred

* Update tests

* Make clarifying comments clearer

* Call asyncio.wait with return_when=ALL_COMPLETED

* Update otbr test

* Update homeassistant/components/thread/dataset_store.py

Co-authored-by: Stefan Agner <stefan@agner.ch>

* Update homeassistant/components/thread/dataset_store.py

---------

Co-authored-by: Stefan Agner <stefan@agner.ch>

Bre77 committed:
Improve coordinator logic in Tessie to allow sleep (#107988)

* Poll status before state

* Tests

bdraco committed:
Small speed up to checking core state (#107845)

edenhaus committed:
Add task to install all requirements of an integration (#108262)

* Add task to install the requirements of an integration

* Gather recursive requirements

* Move valid_integration to util

* Apply suggestions from code review

Co-authored-by: Marc Mueller <30130371+cdce8p@users.noreply.github.com>

* Implement suggestions

---------

Co-authored-by: Marc Mueller <30130371+cdce8p@users.noreply.github.com>

cdce8p committed:
Improve august typing (1) (#108325)

cdce8p committed:
Improve august typing (2) (#108327)

SteveEasley committed:
Fix remote control codes for jvc_projector (#108253)

Update dependency to add/fix remote codes

cdce8p committed:
Improve api typing (#108307)

chatziko committed:
Add iBeacon UUID allowlist (#104790)

cdce8p committed:
Improve august typing (3) (#108329)

cdce8p committed:
Improve august typing (4) (#108331)

cdce8p committed:
Improve august typing (5) (#108332)

cdce8p committed:
Improve person typing (#108218)

cdce8p committed:
Adjust require_admin decorator typing (#108306)

Co-authored-by: J. Nick Koston <nick@koston.org>

cdce8p committed:
Improve ffmpeg* typing (#108092)

bouwew committed:
Bump plugwise to v0.36.3 (#108347)

cereal2nd committed:
Bump pyDuotecno to 2024.1.2 (#108314)

bdraco committed:
Bump orjson to 3.9.12 (#108350)

emontnemery committed:
Fix homekit_controller test (#108375)

bdraco committed:
Avoid json default fallback for area registry (#108358)

cdce8p committed:
Enable strict typing for bluetooth_adapters (#108365)

cdce8p committed:
Enable strict typing for api (#108363)

cdce8p committed:
Enable strict typing for default_config (#108366)

cdce8p committed:
Enable strict typing for my (#108369)

cdce8p committed:
Enable strict typing for map (#108368)

cdce8p committed:
Enable strict typing for intent_script (#108367)

bdraco committed:
Avoid json encoder default fallback for APIComponentsView (#108359)

cdce8p committed:
Improve ifttt typing (#108308)

bdraco committed:
Use shorthand attributes for mobile_app sensor platforms (#108353)

autinerd committed:
Bump openwebifpy to 4.2.1 (#107894)

frenck committed:
Add icon to entity registry list for display (#108313)

eifinger committed:
Expose TimeoutError during google_travel_time config_flow (#108179)

Expose TimeoutError during config_flow

cdce8p committed:
Update dwdwfsapi to 1.0.7 (#108377)

RJPoelstra committed:
Add binary sensor to MotionMount integration (#107659)

* Add binary sensor for `isMoving`

* Sort platforms alphabetically

* Update doc strings

cdce8p committed:
Update types packages (#108371)

edenhaus committed:
Add support for MQTT based ecovacs vacuums (#108167)

* Add support for MQTT based ecovacs vacuums

* renames

* Add init import test

* bump deebot-client

* Translate continent options

* Apply suggestions from code review

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* Remove continent

* use ServiceValidationError

* Small refactoring

* Simplify

* Fix tests

* Enable strict typing for ecovacs

* Apply suggestions from code review

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

* Implement suggestions

* improve test_async_setup_import

* Implement suggestions

* Update homeassistant/components/ecovacs/config_flow.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

---------

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>
Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

frenck committed:
Add icon translations support (#103294)

Co-authored-by: Robert Resch <robert@resch.dev>
Co-authored-by: Paul Bottein <paul.bottein@gmail.com>

cdce8p committed:
Update coverage to 7.4.0 (#108370)

cdce8p committed:
Improve tplink_lte typing (#108393)

cdce8p committed:
Update boto3 to 1.33.13 and aiobotocore to 2.9.1 (#108384)

joostlek committed:
Add icon translations to Anova (#108399)

joostlek committed:
Add icon translations to AirQ (#108402)

joostlek committed:
Add icon translations to Aurora (#108410)

joostlek committed:
Add icon translations to August (#108396)

joostlek committed:
Add icon translations to Aussie Broadband (#108409)

joostlek committed:
Add icon translations to awair (#108408)

joostlek committed:
Add icon translations to Airnow (#108403)

joostlek committed:
Add icon translations to AsusWRT (#108397)

joostlek committed:
Add icon translations to Airthings BLE (#108401)

joostlek committed:
Add icon translations to Ambient station (#108400)

joostlek committed:
Add icon translations to Aseko (#108398)

StevenLooman committed:
Bump async-upnp-client to 0.38.1 (#108382)

autinerd committed:
Use async_create_clientsession for enigma2 (#108395)

Use async_creeate_clientsession for enigma2

frenck committed:
Add alarm_control_panel icon translations (#108413)

* Add alarm_control_panel icon translations

* Nest services correctly

frenck committed:
Add button icon translations (#108415)

frenck committed:
Add calendar icon translations (#108416)

frenck committed:
Update psutil to 5.9.8 (#108421)

frenck committed:
Add air_quality icon translations (#108420)

bdraco committed:
Bump thermopro-ble to 0.8.0 (#108319)

bdraco committed:
Avoid json encoder default fallback when serializing config (#108360)

Co-authored-by: Paulus Schoutsen <balloob@gmail.com>

frenck committed:
Add light icon translations (#108414)

frenck committed:
Mark flaky fritz update test as xfail (#108447)

frenck committed:
Add climate icon translations (#108418)

frenck committed:
Add date icon translations (#108448)

frenck committed:
Add datetime icon translations (#108449)

frenck committed:
Add sensor icon translations (#108450)

* Add sensor icon translations

* Add missing moisture

frenck committed:
Add image icon translations (#108455)

frenck committed:
Add number icon translations (#108452)

frenck committed:
Add event icon translations (#108453)

frenck committed:
Add scene icon translations (#108456)

jrieger committed:
Add icon translations to AVM FRITZ!Box Call Monitor (#108417)

* Add icon translations to AVM FRITZ!Box Call Monitor

* Update homeassistant/components/fritzbox_callmonitor/icons.json

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

* Update homeassistant/components/fritzbox_callmonitor/icons.json

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

---------

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

frenck committed:
Add camera icon translations (#108419)

Co-authored-by: Paulus Schoutsen <balloob@gmail.com>

frenck committed:
Add lock icon translations (#108467)

frenck committed:
Ensure pre-commit runs hassfest when icons change (#108470)

frenck committed:
Add text icon translations (#108457)

frenck committed:
Add humidifier icon translations (#108465)

frenck committed:
Add geo_location icon translations (#108463)

frenck committed:
Add fan icon translations (#108461)

emontnemery committed:
Fix empty files included by !include_dir_named (#108489)

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

jpbede committed:
Use right state class for volume and timestamp sensor in bthome (#107675)

frenck committed:
Add siren icon translations (#108473)

jrieger committed:
Add icon translations to Tankerkoenig (#108499)

frenck committed:
Pin pandas to 2.1.4 (#108509)

farmio committed:
Update knx-frontend to 2024.1.20.105944 (#108511)

elupus committed:
Upgrade nibe to 2.7.0 (#108507)

frenck committed:
Add icon translations to HomeWizard Energy (#108506)

* Add icon translations to HomeWizard Energy

* Update snapshots

frenck committed:
Add icon translations to Plugwise (#108498)

frenck committed:
Add water heater icon translations (#108491)

frenck committed:
Add weather icon translations (#108488)

frenck committed:
Add media player icon translations (#108486)

frenck committed:
Add wake word icon translations (#108482)

frenck committed:
Add update icon translations (#108481)

* Add update icon translations

* Oops

frenck committed:
Add valve icon translations (#108480)

frenck committed:
Add vacuum icon translations (#108479)

Ernst79 committed:
Bump bthome-ble to 3.5.0 (#108475)

frenck committed:
Add tts icon translations (#108476)

frenck committed:
Add todo icon translations (#108477)

frenck committed:
Add select icon translations (#108472)

frenck committed:
Add stt icon translations (#108474)

frenck committed:
Add remote icon translations (#108469)

frenck committed:
Add lawn mower icon translations (#108466)

frenck committed:
Add cover icon translations (#108460)

* Add cover icon translations

* States -> state

frenck committed:
Fix hassfest icon schema for service only (#108494)

emontnemery committed:
Remove unused TypeVar from config.py (#108495)

frenck committed:
Add time icon translations (#108458)

jrieger committed:
Change calendar icon based on state (#108451)

jrieger committed:
Add sun icon translations (#108462)

jrieger committed:
Add nut icon translations (#108471)

bdraco committed:
Remove OrderedDict from auth_store (#108546)

normal dicts keep track of insert order now so this should
no longer be needed

floriankisser committed:
Fix zha illuminance measured value mapping (#108547)

mib1185 committed:
Add icon translations for Pegelonline (#108554)

add icon translations

bdraco committed:
Always load auth storage at startup (#108543)

frenck committed:
Ensure icon translations aren't the same as the default (#108568)

bdraco committed:
Ensure button platform does not restore unavailable state (#108316)

lellky committed:
Add sensors to the flexit_bacnet integration (#108297)

* Adds sensors to the flexit_bacnet integration

* Add one platform at a time

* Removes commented out code

And restores attributes that are needed

* Review changes

* More review fixes

* Adds translations for the flexit_bacnet sensors

* Review comments

* Adds test for flexit_bacnet sensor

* Refactors the sensor test

* Review comment

* Review comment

* Review comments

lellky committed:
Add binary sensors to flexit_bacnet integration (#108571)

* Adds binary sensors to flexit_bacnet integration

* Review comments

* Removes binary sensor for electric heater

Will add switch or service later

zweckj committed:
Add icon translations for lamarzocco (#108569)

icon translations

jrieger committed:
Add icon translation to Jellyfin (#108559)

jrieger committed:
Add icon translations to Roborock (#108508)

jrieger committed:
Add icon translations to DWD Weather Warnings (#108501)

jpbede committed:
Make remaining WLED entities translatable (#108534)

exxamalte committed:
Add unique id to geonetnz_volcano sensors (#108556)

add unique id to each sensor

jbouwh committed:
Icon translation for imap mail count sensor (#108576)

sdb9696 committed:
Add authentication to tplink integration for newer devices (#105143)

* Add authentication flows to tplink integration to enable newer device protocol support

* Add websession passing to tplink integration discover methods

* Use SmartDevice.connect()

* Update to use DeviceConfig

* Use credential hashes

* Bump python-kasa to 0.6.0.dev0

* Fix tests and address review comments

* Add autodetection for L530, P110, and L900

This adds mac address prefixes for the devices I have.
The wildcards are left quite lax assuming different series may share the same prefix.

* Bump tplink to 0.6.0.dev1

* Add config flow tests

* Use short_mac if alias is None and try legacy connect on discovery timeout

* Add config_flow tests

* Add init tests

* Migrate to aiohttp

* add some more ouis

* final

* ip change fix

* add fixmes

* fix O(n) searching

* fix O(n) searching

* move code that cannot fail outside of try block

* fix missing reauth_successful string

* add doc strings, cleanups

* error message by password

* dry

* adjust discovery timeout

* integration discovery already formats mac

* tweaks

* cleanups

* cleanups

* Update post review and fix broken tests

* Fix TODOs and FIXMEs in test_config_flow

* Add pragma no cover

* bump, apply suggestions

* remove no cover

* use iden check

* Apply suggestions from code review

* Fix branched test and update integration title

* legacy typing

* Update homeassistant/components/tplink/__init__.py

* lint

* Remove more unused consts

* Update test docstrings

* Add sdb9696 to tplink codeowners

* Update docstring on test for invalid DeviceConfig

* Update test stored credentials test

---------

Co-authored-by: Teemu Rytilahti <tpr@iki.fi>
Co-authored-by: J. Nick Koston <nick@koston.org>

emontnemery committed:
Fix numbered list in github config flow (#108587)

bdraco committed:
Add bthome event platform (#108268)

Co-authored-by: Ernst Klamer <e.klamer@gmail.com>

lellky committed:
Add missing property in flexit bacnet test (#108606)

Add missing property on mocked device

Also update the snapshot

jpbede committed:
Add icon translations to co2signal (#108611)

bdraco committed:
Use more shorthand attributes in tplink (#108284)

* Use more shorthand attributes in tplink

* naming

* unused

joostlek committed:
Add icon translations to Withings (#108385)

* Add icon translations to Withings

* Add icon translations to Withings

* Add icon translations to Withings

* Add icon translations to Withings

* Add icon translations to Withings

LaStrada committed:
Bump airthings-ble to 0.6.0 (#108612)

zxdavb committed:
Tweak evohome code quality (#107596)

* initial commit

* lint

* initial commit

emontnemery committed:
Fix FlowHandler show progress (#108586)

boralyl committed:
Send recurrence data when updating a task in todoist (#108269)

* Send recurrence data when updating a task in todoist

* Update tests/components/todoist/test_todo.py

Co-authored-by: Allen Porter <allen.porter@gmail.com>

* Move logic into _task_api_data.

* Add comment about sending potentinally stale data.

---------

Co-authored-by: Allen Porter <allen.porter@gmail.com>

codyc1515 committed:
Update August diagnostics.py to redact contentToken (#108626)

bdraco committed:
Significantly reduce websocket api connection auth phase latency (#108564)

* Significantly reduce websocket api connection auth phase latancy

Since the auth phase has exclusive control over the websocket
until ActiveConnection is created, we can bypass the queue and
send messages right away. This reduces the latancy and reconnect
time since we do not have to wait for the background processing
of the queue to send the auth ok message.

* only start the writer queue after auth is successful

bdraco committed:
Speed up run time of admin services by using HassJob (#108623)

bdraco committed:
Avoid recreating persistent notification update function when subscribing (#108624)

I recently came up with an idea to look for callback functions
that get created over and over frequently by adding logging to
homeassistant.core.callback when its called to decorate
a function. This one happens a lot at runtime.

bdraco committed:
Small performance improvements to handing revoke token callbacks (#108625)

- Use a set to avoid linear search for remove
- Avoid recreating the unregister function each time

bdraco committed:
Refactor async_listen_once to remove nonlocal (#108627)

bdraco committed:
Avoid string decode/encode round trip in websocket_api get_services (#108632)

The cache was converting from bytes to str and when we read
the cache we converted it back to bytes again

bdraco committed:
Small cleanups to async_get_all_descriptions (#108633)

bdraco committed:
Bump habluetooth to 2.3.1 (#108628)

bdraco committed:
Small cleanup to zone async_active_zone (#108629)

bdraco committed:
Fix unifi test_tracked_clients test (#108638)

Fix unifi test_tracked_clients client

This test relied on the sensor platform getting set up and creating
the device before the device_tracker platform was setup since the
device_tracker platform will disable the entity because there is
not matching device entry for it via

https://github.com/home-assistant/core/blob/250de165b75a4b4e8e101b156a35b46f9e9c40d7/homeassistant/components/device_tracker/config_entry.py#L336

There is no guarantee the sensor platform will get set up before
the device tracker platform so the test was subject to a race
where it would fail if the device tracker platform was setup first

ludeeus committed:
Remove numbering from GitHub progress step to not deal with styling of list entries (#108639)

jpbede committed:
Remove obsolete services.yaml and translations from WLED (#108605)

jpbede committed:
Add icon translations to WLED (#108604)

joostlek committed:
Add icon translations to Abode (#108407)

rytilahti committed:
Bump songpal dependency to 0.16.1 (#108637)

joostlek committed:
Use default icon in Agent DVR (#108405)

joostlek committed:
Add icon translations to Adguard (#108406)

bdraco committed:
Reduce lock contention when all translations are already cached (#108634)

cdce8p committed:
Update python-slugify to 8.0.1 (#108373)

edenhaus committed:
Add binary_sensor to Ecovacs (#108544)

frenck committed:
Update Pillow to 10.2.0 (#108422)

chemelli74 committed:
Bump aiovodafone to 0.5.4 (#108592)

emontnemery committed:
Convert AreaEntry to dataclass (#108648)

* Convert AreaEntry to dataclass

* Correct typing of AreaEntry.id

* Move responsibility for generating area id to AreaRegistry

jmwaldrip committed:
Bump asyncsleepiq to 1.5.2 (#108431)

Upgrading asyncsleepiq to version 1.5.2

joostlek committed:
Add device to Lutron (#107467)

* Add typing to Lutron platforms

* Add devices to Lutron

* Add devices to Lutron

* Fix typing

* Fix

* Add name

* Fix lights

* Comment out ESA

* Fix domain

* Fix domain

* Fix

* Make generic keypad base class

MartinHjelmare committed:
Add cloud tts entity (#108293)

* Add cloud tts entity

* Test test_login_view_missing_entity

* Fix pipeline iteration for migration

* Update tests

* Make migration more strict

* Fix docstring

jpbede committed:
Add helper function to update and reload config entry to config flow (#108034)

* Add helper function to update and reload config entry to config flow

* Use async_create_task

* Remove await

* Reload only when update & add task name

* Rename function

jpbede committed:
Cleanup Discovergy config flow (#108381)

* Cleanup Discovergy config flow

* Make use of the helper function

bgoncal committed:
Update iOS configuration adding Action toggles to show in CarPlay and Watch (#108355)

frenck committed:
Move core fundamental components into bootstrap (#105560)

Co-authored-by: Erik <erik@montnemery.com>
Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

mib1185 committed:
Prevent runtime issue during entity registration in coordinator of AVM Fritz!Tools (#108667)

prevent dictionary changed size during iteration

Thomas55555 committed:
tests

Thomas55555 committed:
tests

joostlek committed:
Use snapshot testing in Airly sensor (#108608)

* Use snapshot testing in Airly sensor

* Apply suggestions from code review

Co-authored-by: Robert Resch <robert@resch.dev>

* Fix tests

---------

Co-authored-by: Robert Resch <robert@resch.dev>

joostlek committed:
Add icon translations to Airly (#108404)

* Add icon translations to Airly

* Fix test

* Fix tests

jmwaldrip committed:
Fix SleepIQ setting FootWarmer timer (#108433)

* Fixing foot warmer timer bug

* Fixing bug where temperature wasnt assigned to number entity causing tests to fail

Thomas55555 committed:
tests

Thomas55555 committed:
tests

Bre77 committed:
Add time to charge sensor to Tessie (#108342)

* Add time to charge and type checking

* Revert drive_state_shift_state change

* Use original name

* Use function instead of lambda

* Update homeassistant/components/tessie/sensor.py

Co-authored-by: Jan-Philipp Benecke <github@bnck.me>

* Fix callback

* Avoid having to test None

* Go back to if

* Use minutes instead of hours

---------

Co-authored-by: Jan-Philipp Benecke <github@bnck.me>

jpbede committed:
Fix flaky sensibo test (#108669)

Thomas55555 committed:
Merge branch 'dev' into automower

Thomas55555 committed:
Revert "Merge branch 'dev' into automower"

This reverts commit 45d8d1fb3b7889d31d1d6f209c489d6a6b4bcb2a, reversing
changes made to bacbc9bcf8fca28b84227164cc3b9e0898c4afa2.

Thomas55555 committed:
Revert "Revert "Merge branch 'dev' into automower""

This reverts commit c1148fee616e53e13baf9b166b746e9aeb211e3d.

Thomas55555 committed:
Revert "Revert "Merge branch 'dev' into automower""

Thomas55555 committed:
Revert "Revert "Merge branch 'dev' into automower""

Thomas55555 committed:
Merge remote-tracking branch 'refs/remotes/origin/automower' into automower
